operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1673 @@
+package lbipam
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/ipam/service/ipallocator"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+	core_v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	client_typed_v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	operatorOption "github.com/cilium/cilium/operator/option"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/k8s"
+	cilium_api_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	cilium_client_v2alpha1 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_labels "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
+	slim_meta_v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/stream"
+)
+
+const (
+	// The condition added to services to indicate if a request for IPs could be satisfied or not
+	ciliumSvcRequestSatisfiedCondition = "io.cilium/lb-ipam-request-satisfied"
+
+	// The annotation LB IPAM will look for when searching for requested IPs
+	ciliumSvcLBIPSAnnotation = "io.cilium/lb-ipam-ips"
+
+	// The string used in the FieldManager field on update options
+	ciliumFieldManager = "cilium-operator-lb-ipam"
+
+	serviceNamespaceLabel = "io.kubernetes.service.namespace"
+	serviceNameLabel      = "io.kubernetes.service.name"
+)
+
+var Cell = cell.Module(
+	"LB-IPAM",
+	cell.Config(defaultLBIPAMConfig),
+	// Provide LBIPAM so instances of it can be used while testing
+	cell.Provide(registerLBIPAM),
+	// Invoke an empty function which takes an LBIPAM to force its construction.
+	cell.Invoke(func(*LBIPAM) {}),
+)
+
+var defaultLBIPAMConfig = LBIPAMConfig{
+	EnableLBIPAM: false,
+}
+
+type LBIPAMConfig struct {
+	EnableLBIPAM bool `mapstructure:"lb-ipam"`
+}
+
+func (cfg LBIPAMConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool("lb-ipam", cfg.EnableLBIPAM, "Enables LB IPAM, which manages IPs for LoadBalancer services internally without cloud providers")
+}
+
+type LBIPAMInitDone chan struct{}
+
+type LBIPAMParams struct {
+	cell.In
+
+	Logger       logrus.FieldLogger
+	LC           hive.Lifecycle
+	ClientSet    k8sClient.Clientset
+	PoolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	SvcResource  resource.Resource[*core_v1.Service]
+	Config       LBIPAMConfig
+	InitDone     LBIPAMInitDone `optional:"true"`
+}
+
+func registerLBIPAM(params LBIPAMParams) *LBIPAM {
```

Comment:
s/register/new/ ?

Reply:
My preference would be that we don't `Provide` struct pointers, but always interfaces (in `interfaces.go`). With interfaces we'll have a single place to document the "service" this cell is providing and leaves the door open for mocking it in an integration test of another component. We can also return multiple interfaces for different aspects of the API (e.g. Read, Write, Manage). And one more benefit is that it allows having separate internal and external interfaces in case the feature is split into multiple packages and the internal packages need to access some internal stuff.

There's also `ProvidePrivate` now that only exports types inside the module and child modules.